### PR TITLE
Replace magic numbers with macros and add explaining comments

### PR DESCRIPTION
**Pull request by AniruddhaKanhere**
Description
-----------
This PR removes the magic numbers from the code and replaces them with macros.
The value to which the macro is set is explained with comments in Doxygen formatting.

Test Steps
-----------

Checklist:
----------

- [ ] I have tested my changes. No regression in existing tests.
- [ ] I have modified and/or added unit-tests to cover the code changes in this Pull Request.

Related Issue
-----------

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by kstribrnAmzn — DISMISSED** [review]

**Comment by kstribrnAmzn** [comment]
I've approved. I'll take it you're using the build checks for your testing?

File: source/core_mqtt.c [hunk 1]
```diff
@@ -2862,7 +2946,7 @@ MQTTStatus_t MQTT_Ping( MQTTContext_t * pContext )
     MQTTFixedBuffer_t localBuffer;
```
Comment (paulbartell) [inline source/core_mqtt.c:2946]:
```suggestion
    uint8_t pingreqPacket[ MQTT_PACKET_PINGREQ_SIZE ];
```

File: source/core_mqtt.c [hunk 2]
```diff
@@ -2877,6 +2961,7 @@ MQTTStatus_t MQTT_Ping( MQTTContext_t * pContext )
```
Comment (paulbartell) [inline source/core_mqtt.c:2961]: Line above isnt necessary.

File: source/core_mqtt.c [hunk 3]
```diff
@@ -2877,6 +2961,7 @@ MQTTStatus_t MQTT_Ping( MQTTContext_t * pContext )
 
         if( status == MQTTSuccess )
         {
+            assert( packetSize == localBuffer.size );
```
Comment (paulbartell) [inline source/core_mqtt.c:2964]:
```suggestion
```

File: source/core_mqtt.c [hunk 4]
```diff
@@ -2904,7 +2989,7 @@ MQTTStatus_t MQTT_Ping( MQTTContext_t * pContext )
          * from the user provided buffers. Thus it can be sent directly. */
         sendResult = sendBuffer( pContext,
                                  localBuffer.pBuffer,
-                                 2U );
+                                 packetSize );
```
Comment (paulbartell) [inline source/core_mqtt.c:2992]:
```suggestion
                                 MQTT_PACKET_PINGREQ_SIZE );
```

File: source/core_mqtt.c [hunk 5]
```diff
@@ -2753,10 +2840,7 @@ MQTTStatus_t MQTT_Publish( MQTTContext_t * pContext,
     size_t packetSize = 0UL;
     MQTTPublishState_t publishStatus = MQTTStateNull;
     bool stateUpdateHookExecuted = false;
-
-    /* 1 header byte + 4 bytes (maximum) required for encoding the length +
-     * 2 bytes for topic string. */
-    uint8_t mqttHeader[ 7 ];
+    uint8_t mqttHeader[ CORE_MQTT_PUBLISH_PACKET_HEADER_MAX_LENGTH ];
```
Comment (paulbartell) [inline source/core_mqtt.c:2843]: IMO, this doesn't really help readability.

File: source/core_mqtt.c [hunk 6]
```diff
@@ -1882,22 +1975,20 @@ static MQTTStatus_t sendSubscribeWithoutCopy( MQTTContext_t * pContext,
                                               size_t remainingLength )
 {
     MQTTStatus_t status = MQTTSuccess;
-    uint8_t subscribeheader[ 7 ];
+    uint8_t subscribeheader[ CORE_MQTT_SUBSCRIBE_PACKET_HEADER_MAX_LENGTH ];
     uint8_t * pIndex;
     TransportOutVector_t pIoVector[ MQTT_SUB_UNSUB_MAX_VECTORS ];
     TransportOutVector_t * pIterator;
-    uint8_t serializedTopicFieldLength[ MQTT_SUB_UNSUB_MAX_VECTORS ][ 2 ];
```
Comment (paulbartell) [inline source/core_mqtt.c:1889]: Seems like a 16-bit length type would be a simpler way to handle this?

File: source/core_mqtt.c [hunk 7]
```diff
@@ -161,7 +256,7 @@ static int32_t sendMessageVector( MQTTContext_t * pContext,
  *
  * @return The number of vectors added.
  */
-static size_t addEncodedStringToVector( uint8_t serailizedLength[ 2 ],
```
Comment (paulbartell) [inline source/core_mqtt.c:164]: A typedef seems like a better choice here.

File: source/core_mqtt.c [hunk 8]
```diff
@@ -1833,29 +1928,27 @@ static MQTTStatus_t validateSubscribeUnsubscribeParams( const MQTTContext_t * pC
...
     TransportOutVector_t * pLocalIterator = iterator;
-    /* This function always adds 2 vectors. */
     size_t vectorsAdded = 0U;
 
     /* When length is non-zero, the string must be non-NULL. */
     assert( ( length != 0U ) == ( string != NULL ) );
 
-    serailizedLength[ 0 ] = ( ( uint8_t ) ( ( length ) >> 8 ) );
-    serailizedLength[ 1 ] = ( ( uint8_t ) ( ( length ) & 0x00ffU ) );
+    serializedLength[ 0 ] = ( ( uint8_t ) ( ( length ) >> 8 ) );
```
Comment (paulbartell) [inline source/core_mqtt.c:1944]: A macro or static inline helper function would improve clarity here

File: source/core_mqtt.c [hunk 9]
```diff
@@ -70,6 +70,101 @@
...
+
+/**
+ * @brief Maximum number of vectors required to encode and send a publish
+ * packet. The breakdown is shown below.
+ * Fixed header (including topic string length)      0 + 1 = 1
+ * Topic string                                        + 1 = 2
+ * Packet ID (only when QoS > QoS0)                    + 1 = 3
+ * Payload                                             + 1 = 4
+ */
+#define CORE_MQTT_PUBLISH_PACKET_MAX_VECTORS              ( 4U )
```
Comment (paulbartell) [inline source/core_mqtt.c:166]: CORE_MQTT_PUBLISH_PACKET_MAX_VECTORS is used in only one place. Moving this comment down there could improve clarity.

File: source/core_mqtt.c [hunk 10]
```diff
@@ -70,6 +70,101 @@
...
+ * @brief The maximum vectors required to encode and send a connect packet. The
+ * breakdown is shown below.
+ * Fixed header      0 + 1 = 1
+ * Client ID           + 2 = 3
+ * Will topic          + 2 = 5
+ * Will payload        + 2 = 7
+ * Username            + 2 = 9
+ * Password            + 2 = 11
+ */
+#define CORE_MQTT_CONNECT_PACKET_MAX_VECTORS              ( 11U )
```
Comment (paulbartell) [inline source/core_mqtt.c:156]: Similar situation here. This is only used once.

File: source/core_mqtt.c [hunk 11]
```diff
@@ -70,6 +70,101 @@
...
+ * Header byte           0 + 1 = 1
+ * Length (max)            + 4 = 5
+ * Topic string length     + 2 = 7
+ *
+ * @note Since publish is one of the most common operations in MQTT connection,
+ * we have moved the topic string length to the 'fixed' part of the header so
+ * efficiency. Otherwise, we would need an extra vector and an extra call to
+ * 'send' (in case writev is not defined) to send the topic length.
+ */
+#define CORE_MQTT_PUBLISH_PACKET_HEADER_MAX_LENGTH        ( 7U )
```
Comment (paulbartell) [inline source/core_mqtt.c:144]: Same deal here.

**Review by paulbartell — CHANGES_REQUESTED** [review]
The additional comments are certainly useful. I don't think they warrant macros though, especially when used in a single place.

**Review by paulbartell — APPROVED** [review]

**Review by kstribrnAmzn — APPROVED** [review]

**Pull request merged**